Feat/add zustand

affan880 opened the pull request:
Refactor authentication flow and API integration
- Introduced a new API utility for handling generic API calls.
- Updated authentication store to manage login, registration, and user details.
- Enhanced routing logic in App component to differentiate between public and protected routes.
- Added user details form validation schema.
- Updated Login and Register pages to utilize the new authentication store and handle navigation based on user authentication status.
- Conditionally showing user details form.

jackli921 committed:
WIP - add zustand

affan880 committed:
Refactor authentication flow and API integration

- Introduced a new API utility for handling generic API calls.
- Updated authentication store to manage login, registration, and user details.
- Enhanced routing logic in App component to differentiate between public and protected routes.
- Added user details form validation schema.
- Updated Login and Register pages to utilize the new authentication store and handle navigation based on user authentication status.
- Adjusted import paths for consistency.
- Conditionally showing user details form.

jackli921 commented on apps/client/src/lib/schema.tsx:
```diff
@@ -30,9 +30,22 @@ export const registerSchema = z
+  default_currency: z
+    .string()
+    .min(1, { message: 'Default currency is required' }),
+  monthly_budget: z.number().min(1, { message: 'Monthly budget is required' }),
```
I thought about whether to get a monthly budget as a single value or calculate the monthly budget by summing the budget given to each category of spending. My conclusion is that we should always calculate the current monthly budget dynamically based on the total of budget given to each category.

Reason:
1. If we allow user to enter a specific monthly budget value, we'd need to add additional logic to ensure that individual budget or total of budget(s)  cannot exceed this value. 
2. In order to maintain data integrity, we'd always need to make additional update calls to update this singular value if individual budget amount changes, which we have not determined how to do.
3. Having a monthly budget doesn't help us calculate the difference between money spent and budget left for individual categories.

So, I think we can leave out this monthly_budget field for now.

affan880 replied:
so should I include the field In the form as of now or should I exclude it?

jackli921 replied:
I think we can exclude it from the form

jackli921 replied:
I think the monthly budget is still included in the form component?

affan880 replied:
yes, I will Update it while implementing API call to update the user details in our db

jackli921 commented on apps/client/src/stores/authStores.ts:
```diff
@@ -0,0 +1,106 @@
+  saveUserDetails: async (data: UserDetailsForm) => {
+    set({ isLoading: true });
+    try {
+      //   const response = await post(AUTH_ENDPOINTS.USER_DETAILS, data, {});
```
what's the reason for commenting this part out?

affan880 replied:
Because the endpoint to update the users detail is not yet created.

jackli921 replied:
Ah ok.

jackli921 commented on apps/client/src/stores/authStores.ts:
```diff
@@ -0,0 +1,106 @@
+    try {
+      //   const response = await post(AUTH_ENDPOINTS.USER_DETAILS, data, {});
+      console.log('######DATA########################', data);
+      set({ defaultCurrency: 'USD', isLoading: false });
```
This is probably related to the previous question, but is this trying to set the default currency of a newly created user to be 'USD'? This `saveUserDetails` call is a call to the backend api to save the user selection, right?

affan880 replied:
Because the endpoint to update the users detail is not yet created. I used dummy value to check the flow util we create user details endpoint

jackli921 reviewed (commented):
good stuff, just have some questions

affan880 committed:
Fix import path for HomePage component in App.tsx

affan880 committed:
Update import paths for consistency in Auth pages and enhance error handling in LoginPage

jackli921 commented on apps/client/src/pages/Auth/LoginPage.tsx:
```diff
@@ -23,9 +28,29 @@ const LoginPage = () => {
+      const response = await authLogin(data);
+      console.log('######RESPONSE########################', response);
+      if (defaultCurrency) {
+        navigate('/Home');
```
lowercase the path?

affan880 replied:
okay

jackli921 commented on apps/client/src/pages/Auth/UserDetailsPage.tsx:
```diff
@@ -0,0 +1,166 @@
+                  className="text-sm text-gray-600"
+                >
+                  Monthly Budget
+                </label>
```
label tag is for default_curency but the text is Monthly Budget?

jackli921 commented on apps/client/src/pages/Auth/UserDetailsPage.tsx:
```diff
@@ -0,0 +1,166 @@
+                  transition-all duration-200 placeholder:text-gray-400 
+                  text-black"
+                  placeholder="Enter your monthly budget"
+                />
```
I think we can delete the monthly budget input like discussed

jackli921 commented on apps/client/src/pages/Auth/LoginPage.tsx:
```diff
@@ -23,9 +28,29 @@ const LoginPage = () => {
-    setError(null);
+    try {
+      if (isAuthenticated) {
+        authLogout();
```
why call authLogout within the onSubmit of a login attempt?

affan880 replied:
if the user is already logged In and or has cached token

jackli921 replied:
if they are already logged in or has cached token, we log them out? I don't follow.

jackli921 commented on apps/client/src/api/axiosConfig.ts:
```diff
@@ -0,0 +1,34 @@
+import axios from 'axios';
+
+const API = axios.create({
+  baseURL: 'http://localhost:3000/api/v1',
```
update this to use environment variables or the default localhost port?

```typescript
baseURL: API_BASE_URL || 'http://localhost:3000/api/v1',
```

kayoMichael replied:
Since we are using vite. We have to call the env variables like this
```
import.meta.env.VITE_API_BASE_URL;
```

affan880 replied:
okay @kayoMichael  thank you for letting me know that. I didn't knew that

jackli921 commented on apps/client/src/api/axiosConfig.ts:
```diff
@@ -0,0 +1,34 @@
+  (error) => {
+    if (error.response && error.response.status === 401) {
+      console.warn('Unauthorized request. Redirecting to login...');
+      localStorage.removeItem('token');
+    }
```
1. If we can get access to the authStore, we should nullify the token stored in the AuthStore here too.
2. redirect user to /login?

affan880 replied:
sorry, I didn't get it

jackli921 replied:
What I meant to say was that should set the state that stores the token to null. But now that I think about it, since client side javascript won't have access to token stored in HttpOnly cookies. There is no token to call `localStorage.removeItem('token') nor setting token to null because we won't have access to the token in the client anyways.

We can redirect the user to the /login page after line 28 when we actually log them out to improve the user experience.

jackli921 commented on apps/client/src/stores/authStores.ts:
```diff
@@ -0,0 +1,106 @@
+      return response;
+    } catch (error: unknown) {
+      set({ isLoading: false, isAuthenticated: false });
+      if (error instanceof Error) {
+        const errorMessage =
+          (error as AxioError).response?.data?.message ||
+          error.message ||
+          'An unexpected error occurred.';
+        throw new Error(errorMessage);
```
could potentially eliminate recreating a errorMessage here and just rethrow the error because the error here already has detail error messages as result of being processed catch of the `apiCall` in api.ts, right?

```typescript
} catch (error: unknown) {
  set({ isLoading: false, isAuthenticated: false });
  throw error; // just rethrow the error here
}
```

jackli921 commented on apps/client/src/api/axiosConfig.ts:
```diff
@@ -0,0 +1,34 @@
+    const token = localStorage.getItem('token');
+    if (token) {
+      config.headers.Authorization = `Bearer ${token}`;
+    }
```
I don't think we store the JWT token in localStorage or session. The JWT token is signed, issued, and sent in the cookie of the response. Cookies are automatically added to the request of the origin that issued the cookie, as httpyOnly setting prevents clientside javascript accessing HttpOnly cookies to prevent XSS attacks. So this is probably not needed?

jackli921 reviewed (changes requested):
Added some more comments. Overall looks good.

kayoMichael commented on apps/client/src/App.tsx:
```diff
@@ -1,19 +1,40 @@
+
+function App() {
+  const { isAuthenticated, checkAuth } = useAuthStore();
+  useEffect(() => {
```
So I haven;t done any research but I assume everything is still client side rendered because you can use `useEffect`

is SSR just not turned on right now. I guess we gotta do research because I know react 19 has SSR

jackli921 replied:
Yeah. Similar thought, just haven't had enough to time get around all the researching ...

jackli921 commented:
A reminder to rebase the new changes merged in development along with the changes suggested before testing it locallly

affan880 committed:
Refactor routing, enhance user details management and fixed the issues raised by the team

- Updated App component to streamline public and protected routes with navigation handling.
- Integrated currency fetching in UserDetailsPage for currency selection.
- Improved error handling and user feedback in LoginPage and UserDetailsPage.
- Enhanced auth store to manage user details and currency updates.
- Removed monthly budget field from user details form validation.

affan880 committed:
Refactor LoginPage and update server workflow

- Fixed the auth flow in conditionally showing user details form

affan880 committed:
Updated user object structure in auth store to include default currency ID.
